File: 2331-evaluate-boolean-binary-tree/solution.py
# https://leetcode.com/problems/evaluate-boolean-binary-tree/description/
# 2331-evaluate-boolean-binary-tree

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def evaluateTree(self, root: Optional[TreeNode]) -> bool:
        return self.solution1(root)

    def solution(self, root: Optional[TreeNode]) -> bool:
        if root is None or 0 <= root.val < 2:
            return root.val is not None and root.val == 1

        leftRes = self.evaluateTree(root.left)
        rightRes = self.evaluateTree(root.right)

        return (leftRes and rightRes) if root.val == 3 else (leftRes or rightRes)
    
    def solution1(self, root: Optional[TreeNode]) -> bool:
        if root is None or 0 <= root.val < 2:
            return root.val is not None and root.val == 1

        leftRes = self.evaluateTree(root.left)
        if (root.val == 2 and leftRes) or (root.val == 3 and not leftRes):
            return leftRes
        
        return self.evaluateTree(root.right)

File: 2331-evaluate-boolean-binary-tree/solution.ts
// https://leetcode.com/problems/evaluate-boolean-binary-tree/
// 2331-evaluate-boolean-binary-tree

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function evaluateTree(root: TreeNode | null): boolean {
  if(!root || root.val === 0) {
    return false
  }
  if(root.val === 1) {
    return true;
  }
  return root.val === 3 ?
    evaluateTree(root.left) && evaluateTree(root.right) :
    evaluateTree(root.left) || evaluateTree(root.right);
};
